skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -725,22 +872,196 @@ def _fast_nl_means_denoising_3d(cnp.ndarray[np_floats, ndim=3] image,
+    )
+
+
+def _fast_nl_means_denoising_4d(cnp.ndarray[np_floats, ndim=5] image,
+                                Py_ssize_t s=3, Py_ssize_t d=3, double h=0.1,
+                                double var=0.):
+    """
+    Perform fast non-local means denoising on 3-D array, with the outer
+    loop on patch shifts in order to reduce the number of operations.
+
+    Parameters
+    ----------
+    image : ndarray
+        4-D input data to be denoised.
+    s : Py_ssize_t, optional
+        Size of patches used for denoising.
+    d : tuple of Py_ssize_t, optional
+        Maximal distance in pixels along each axis to search for patches used for denoising.
+    h : double, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
+        one is in accepting patches.
+    var : double
+        Expected noise variance.  If non-zero, this is used to reduce the
+        apparent patch distances by the expected distance due to the noise.
+
+    Returns
+    -------
+    result : ndarray
+        Denoised image, of same shape as input image.
+
+    References
+    ----------
+    J. Darbon, A. Cunha, T.F. Chan, S. Osher, and G.J. Jensen, Fast
+    nonlocal filtering applied to electron cryomicroscopy, in 5th IEEE
+    International Symposium on Biomedical Imaging: From Nano to Macro,
+    2008, pp. 1331-1334.
+
+    Jacques Froment. Parameter-Free Fast Pixelwise Non-Local Means
+    Denoising. Image Processing On Line, 2014, vol. 4, p. 300-326.
+    """
+
+    cdef double DISTANCE_CUTOFF = 5.0
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+
+    if np_floats is cnp.float32_t:
+        dtype = np.float32
+    else:
+        dtype = np.float64
+
+    cdef Py_ssize_t offset = s / 2
+    # Image padding: we need to account for patch size, possible shift,
+    # + 1 for the boundary effects in finite differences
+    cdef Py_ssize_t pad_size = offset + d + 1
+    cdef double [:, :, :, :, ::1] padded = np.ascontiguousarray(
+        np.pad(image,
+               ((pad_size, pad_size),
+                (pad_size, pad_size),
+                (pad_size, pad_size),
+                (pad_size, pad_size),
+                (0, 0)),
+               mode='reflect'),
+        dtype=np.float64)
+    cdef double [:, :, :, :, ::1] result = np.zeros_like(padded)
+    cdef double [:, :, :, ::1] weights = np.zeros_like(padded[..., 0])
+    cdef double [:, :, :, ::1] integral = np.zeros_like(padded[..., 0])
+    cdef Py_ssize_t n_pln, n_row, n_col, t_pln, t_row, t_col, \
+             pln, row, col, channel, n_channels, t_time, n_time, time
+    cdef Py_ssize_t time_dist_min, time_dist_max, pln_dist_min, pln_dist_max, \
+             row_dist_min, row_dist_max, col_dist_min, col_dist_max,
+    cdef Py_ssize_t d_row, d_col, d_pln, d_time
+    cdef double weight, distance, alpha
+    n_time, n_pln, n_row, n_col, n_channels = padded.shape[0], padded.shape[1], padded.shape[2], padded.shape[3], padded.shape[4]
```

Comment:
```suggestion
    n_time, n_pln, n_row, n_col, n_channels = padded.shape
```
Does tuple unpacking work here?

Reply:
Apparently not!:
```
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
                 pln, row, col, channel, n_channels, t_time, n_time, time
        cdef Py_ssize_t time_dist_min, time_dist_max, pln_dist_min, pln_dist_max, \
                 row_dist_min, row_dist_max, col_dist_min, col_dist_max,
        cdef Py_ssize_t d_row, d_col, d_pln, d_time
        cdef double weight, distance, alpha
        n_time, n_pln, n_row, n_col, n_channels = padded.shape
                                                       ^
    ------------------------------------------------------------

    skimage/restoration/_nl_means_denoising.pyx:970:52: Assignment to/from slice of wrong length, expected 8, got 5
```


elsewhere where image is a `cnp.ndarray` I get a different error instead if I use this style:
```
        cdef Py_ssize_t n_pln, n_row, n_col
        n_pln, n_row, n_col = image.shape
                                  ^
    ------------------------------------------------------------

    skimage/restoration/_nl_means_denoising.pyx:251:31: Cannot convert 'npy_intp *' to Python object
```